jib-core/src/test/java/com/google/cloud/tools/jib/http/TestWebServer.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2018 Google LLC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.http;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+
+/** Simple local web server for testing. */
+class TestWebServer implements Closeable {
+
+  private final ServerSocket serverSocket;
+  private final ExecutorService executorService = Executors.newSingleThreadExecutor();
+  private final Semaphore threadStarted = new Semaphore(0);
+
+  TestWebServer() throws IOException, InterruptedException {
+    serverSocket = new ServerSocket(0);
+    executorService.submit(this::serve200);
+    threadStarted.acquire();
+  }
+
+  String getEndpoint() {
+    String host = serverSocket.getInetAddress().getHostAddress();
+    return "http://" + host + ":" + serverSocket.getLocalPort();
+  }
+
+  @Override
+  public void close() throws IOException {
+    serverSocket.close();
+    executorService.shutdown();
```

Comment:
Should this go before `serverSocket.close()`? I'm thinking if the `serverSocket.close()` is called while line 56 is running, it would throw an exception.

Reply:
That's the point of `serverSocket.close()`, because `accept()` is a blocking call; that is, to terminate the thread.

Also, calling `executorService.shutdown()` doesn't help. It just prevents future scheduling. Generally, it's not possible to kill a thread. `shutdownNow()` can't do it either.

Reply:
FYI, the main reason for `shutdown()` is related to terminating JVM: https://stackoverflow.com/questions/29999109/executorservice-jvm-doesnt-terminate